contracts/Pool.sol

```diff
@@ -1,31 +1,292 @@
+    /// gain access to funds collected from all blocks from `B - cycleBlocks` to `B - 1`.
+    uint64 internal immutable cycleBlocks;
+    /// @notice Block number on which all payments must be finished
+    uint64 internal immutable maxBlockNumber;
+    /// @notice Maximum sum of all receiver weights of a single sender.
+    /// Limits loss of per-block payments accuracy, they are always multiples of weights sum.
+    uint32 internal constant SENDER_WEIGHTS_SUM_MAX = 1000;
+    /// @notice Maximum number of receivers of a single sender.
+    /// Limits costs of changes in sender's configuration.
+    uint32 internal constant SENDER_WEIGHTS_COUNT_MAX = 100;
+
     struct Sender {
-        uint256 balance;
+        /// @notice Block number on which funding has been started
+        uint64 startBlock;
+        /// @notice The amount available when funding has been started
+        uint192 startBalance;
+        // --- SLOT BOUNDARY
+        /// @notice The target amount sent on each block.
+        /// The actual amount is rounded down to the closes multiple of `weightSum`.
+        uint192 amtPerBlock;
+        /// @notice The total weight of all the receivers
+        uint32 weightSum;
+        /// @notice The number of the receivers
+        uint32 weightCount;
+        // --- SLOT BOUNDARY
+        /// @notice The mapping of all the receivers to their weights, iterable
+        mapping(address => ReceiverWeight) receiverWeights;
+    }
+
+    struct Receiver {
+        /// @notice The block until which funds have been collected previously
+        uint64 lastCollectedCycle;
+        /// @notice The amount which has been previously collected
+        uint192 lastCollectedAmt;
+        // --- SLOT BOUNDARY
+        /// @notice The changes of collected amounts on specific cycle.
+        /// The keys are cycles, each cycle becomes collectable on block `C * cycleBlocks`
+        mapping(uint64 => int256) amtDeltas;
     }
 
+    /// @notice Details about all the senders, the key is the owner's address
     mapping(address => Sender) internal senders;
+    /// @notice Details about all the receivers, the key is the owner's address
+    mapping(address => Receiver) internal receivers;
+
+    /// @param _cycleBlocks The value of cycleBlocks to be used in the instance.
+    /// Low values make funds more available by shortening the average duration of funds being
+    /// frozen between being taken from senders' balances and being collectable by the receiver.
+    /// High values make collecting cheaper by making it process less cycles for a given time range.
+    constructor(uint64 _cycleBlocks) public {
+        cycleBlocks = _cycleBlocks;
+        /// @notice The block number of the last block which is described by the last cycle,
+        /// which is applied to receivers in block which fits in uint64
+        maxBlockNumber = (type(uint64).max / _cycleBlocks - 1) * _cycleBlocks;
```

Comment:
I know, sounds crazy :D But it's necessary to cap some calculations, e.g. we're calculating the moment when the payment will stop. It's easy to top up your account with 2^64 wei (~6600 USD) and set payment rate to 1 wei per block, which if uncapped would mess up our calculations based on assumption that every block number can be stored in a 64-bit number.

Reply:
> in 500 million years.

as a father, I can only tell you that time flies!

Reply:
Thanks for the explanation @CodeSandwich. If I understand this correctly we want to ensure that `blockEnd` can be represented in 64 bit and that `cycle1` in `setSingleDelta` does not overflow. To simplify things we could just cap the `blockEnd` at `2^64 - 2` and revert the transaction with an error if the value is larger. Might make this a bit easier to understand.

Reply:
You're right, your solution works just as good while being simpler! It doesn't guarantee that in `setSingleDelta` `cycle2` will store a cycle translatable to a 64-bit block number, but that's fine, because that won't be needed for another 500 million years.

Reply:
Fixed